ESPS/general/src/plot_script/plotsd.1
.\" Copyright (c) 1987,1988,1989, 1990 Entropic Speech, Inc.; All rights reserved
.\" @(#)plotsd.1	3.22	03 Apr 1997	ESI
.TH PLOTSD 1\-ESPS 03 Apr 1997
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH NAME
plotsd \- plot real sampled data from single-channel FEA_SD files
.SH SYNOPSIS
.B plotsd
[
.BI \-{prs} " range"
] [
.BI \-t " text"
] . . .  [
.BI \-x " level"
] [
.BI \-y " range"
] [
.BI \-E " expansion"
]
.br
[
.BI \-M " maxpnts"
] [
.BI \-T " device"
] [
.BI \-Y " range"
] [
.BI \-W " generic window system options"
]
.I file
.SH DESCRIPTION
.PP
.I Plotsd
plots a portion of an ESPS sampled data file
.I file.   
The file must be a single-channel file.  If the file is complex, only
the real part is plotted.  For multi-channel or complex files,
consider using \fIdemux\fP (1\-\s-1ESPS\s+1) together with
\fIplotsd\fP, or \fIgenplot\fP (1\-\s-1ESPS\s+1) to plot directly.
.PP
If
.I file
is "\-", the standard input is read.
By default,
the output is in Unix GPS (``Graphics Primitive String'') format
and is displayed on the Masscomp or Sun display (using \fImcd\fR), 
but with the
.B \-T
option, this may be sent to the standard output,
obtained in Tektronix format rather than GPS,
or plotted on a hardcopy device.
.PP

.PP
If the \fB\-p\fP option is omitted, then the \fIrange\fP is computed
from the parameter and common files; otherwise the entire contents
of 
.I file 
is used as the default range.  
.PP
.I Plotsd
finds the maximum and  minimum values in the specified range 
and computes appropriate scales for the 
.IR x -
and 
.IR y -axes.
These automatic values can be overriden by the 
.B \-y
or
.B \-Y 
options.  
.PP
When displayed on the screen (see EXAMPLES, below), the plots
will take a long time if every point is plotted.  For this reason, 
.I plotsd
plots a maximum of approximately 3000 points across the range 
specified by the user (intermediate points are skipped).  This behavior
can be changed by means of the 
.B \-M 
option. 
.PP
.I Plotsd
will run faster if the 
.B \-Y
option is used, especially in the case of large files.  
.PP
On Masscomp workstations the color and line style of output 
to ``gps'' or ``mcd'' depend on the
environment variable BUNDLE.  The default is GPS bundle number 20, which
gives solid lines on monochrome displays.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-p " first:last"
.TP
.BI \-p " first:+incr
.TP
.BI \-r " first:last"
.TP
.BI \-r " first:+incr
Determines the range of points to be plotted.  In the first form
(first:last), a pair of unsigned integers gives the first and last
points of the range.  If
.I first
is omitted, 1 is used.  If 
.I last 
is omitted, the last point in the file is used.  The second form
(first:+incr) is equivalent to the first with
.I "last = first + incr".  
The \fB-r\fP is a synonym for \fB-p\fP.  
.TP
.BI \-s " range"
Select a range of points to be plotted in seconds. For compatibility with
DSC LISTEN, the beginning of the file is 0.
.TP
.BI \-t " text"
A line of text to be printed below the plot.  
Up to ten
.B \-t
options may be specified; they are printed as text lines in order below the
plot.
.TP
.BI \-x " level"
If
.I level
is nonzero, debugging information is written to the standard error output.
Default is 0 (no debugging output).
.TP
.BI \-y " low:high"
Specifies approximate lower and upper limits for the 
.IR y -axis.
The values are adjusted by 
.I plotsd
in an attempt to make an esthetically pleasing plot.  
.TP
.BI \-E " expansion"
The range selected by the
.B \-p
option or by the parameter and common files is shown in a context
consisting of additional points at each end.
The selected range is expanded by a fraction given by the argument
.I expansion
plus an additional amount chosen to make the endpoints round numbers.
The expanded range is plotted, and the ends of the actually chosen range are
flagged with vertical lines labeled with the
.IR x- coordinate
values.
.TP
.BI \-M " maxpnts\fR [3000]\fP"
Specifies that 
.I plotsd
plot up to an approximate maximum of 
.I maxpnts 
points.  If the total number of points in the 
.I range
exceeds 
.I maxpnts,
points are skipped evenly throughout.
As a special case,
.I maxpnts
= 0 specifies that no points are to be skipped.
.TP
.BI \-T " device"
.I Device
may be
.BR gps ,
.BR mcd ,
.BR tek ,
.BR suntools ,
or
.BR hardcopy .
The default is
.BR \-Tmcd,
which displays the output on the Sun or Masscomp screen by piping it through
.I mcd 
(on Masscomp systems \fImcd\fR is supplied by Masscomp, on Suns,
it is supplied by ESI).
Specifying
.B \-Tgps
sends output to the standard output in Masscomp Unix GPS format.
Specifying
.B \-Ttek
causes output to be sent to the standard output in Tektronix 4000-series
format.
If
.B \-Tsuntools
is used, then the display will be forced to use \fIsuntools\fR window
system (on Suns and Solbournes only), even if ESPS has been configured
to use the X window system.   If ESPS is built for X, then it uses X for
plotting by default.
If
.B \-Tmcd
is used on a Masscomp and ESPS has been configured to use X windows,
then the old SP-40 \fImcd\fR is used instead of X windows.
Specifying
.B \-Thardcopy
plots it on a hardcopy device, if one was set up in the installation
script.  Note that the word \fIimagen\fR was used for this option in
previous versions.   It is still accepted, but \fIhardcopy\fR is meant
to be more general.
.TP
.BI \-Y " low:high"
Specifies exact lower and upper limits for the 
.IR y -axis.
.TP
.BI \-W
This option is followed by a character
string consisting of the generic window system options.   These options are
passed onto \fImcd\fR or \fIxmcd\fR.   
This option is used to pass along generic X window
arguments to \fIxmcd\fR or Suntool arguments to \fImcd\fR on Suns.  This
option might not have an effect on all systems.   For example, on a
system running X windows, to cause the plot to appear at a particular
screen location the following command could be used: 
.br
\fIplotsd -W "=300x250+50+200" file\fR.
.br
See \fImcd(1\-ESPS)\fR and \fIxmcd(1\-ESPS)\fR for details.
.pp
If the file to be plotted contains data of a single value, then the
horizontal grid is suppressed.  Otherwise the horizontal plot tends to
get lost on top of a grid line.
.SH EXAMPLE
.IP
.BI "plotsd " "<options> <filename>"
.PP
is equivalent to
.IP
.BI "plotsd \-Tgps " "<options> <filename> " "| mcd"
.PP
The second form may be used if nondefault options for
.I mcd
are desired.
.PP
For Sun systems, this program will plot under either X windows, or
Suntools.   By default, if the \fB-T\fR option is not used, it will 
expect X windows.   To plot under Suntools you can use the \fB-Tsuntools\fR
option.    If you always use Suntools, you can avoid using the \fB-T\fR
option by defining the environment variable \fBEPLOTDEV\fR to be
\fIsuntools\fR.  This will make the Suntools window system the default.
This variable is ignored if the machine type is not Sun4, SUN3, or
SUN386i.
.SH "ESPS PARAMETERS"
No ESPS parameter file is read.
.SH "ESPS COMMON"
The following items are read from the ESPS Common file.  The values for
start and nan are not used if a range is specified on the command line or
if the input file is not the same file identified in the
Common file by the "filename" symbol. 
.IP
.I "start - integer"
.IP
This is the starting point in the sampled data file to begin
plotting. If it is zero, the beginning of the file is used.
.sp
.I "nan - integer"
.IP
This is the number of points (starting at 
.IR start )
to plot.
If it is zero, the entire file is used as the default.  
.sp
.I "filename - string"
.IP
This is the name of the file to be displayed if \fIfile\fP is omitted
from the command line.
.PP
The following items are written into the ESPS Common file:
.IP
.I "start - integer"
.IP
The left point in the range selected.
.sp
.I "nan - integer"
.IP
The number of points in the selected range.
.sp
.I "beginplot - integer"
.IP
The origin of the graph
.RI ( x -coordinate).
.sp
.I "endplot - integer"
.IP
The end point of the graph
.RI ( x -coordinate).
.sp
.I "prog - string"
.IP
This is the name of the program (\fIplotsd\fP in this case).
.sp
.I "filename - string"
.IP
This is the name of the file to be displayed.
The program terminates if \fIfilename\fP does not exist when \fIfile\fP
is not specified on the command line.
.PP 
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.SH "ESPS HEADERS"
.PP
.I Plotsd
reads the following field from the SD file header:
.I common.type.
.SH DIAGNOSTICS
.PP
.I Plotsd
complains if there is no such file, if the file is not a sampled data
file, or if the file is a mutli-channel file.  
.SH WARNING
Discarding points when more than
.I maxpoints
points are to be plotted can produce misleading results.
In particular, aliasing is possible if the data has periodic components
with periods comparable to or less than the spacing between the selected
points.
.SH EXPECTED CHANGES
.PP
Many more options may be added.
.SH SEE ALSO
\fIdemux\fP (1\-\s-1ESPS\s+1), \fImlplot\fP (1\-\s-1ESPS\s+1),
\fIgenplot\fP (1\-\s-1ESPS\s+1), 
.br
\fIscatplot\fP (1\-\s-1ESPS\s+1),
\fIplotspec\fP (1\-\s-1ESPS\s+1), \fIaplot\fP (1\-\s-1ESPS\s+1),
.br
FEA_SD (5\-\s-1ESPS\s+1)
.SH AUTHOR
.PP
.nf
Joseph T. Buck, Entropic Speech, Inc.
Adapted from a program by Shankar Narayan.
Converted by Rod Johnson to read new ESPS format.
Modified for ESPS Common by Ajaipal S. Virdy.
Modified for \-y, \-Y, and \-M by John Shore.
Manual page revised by John Shore.
\-T and \-E options added by Rod Johnson.
SunView changes by Alan Parker
.fi


ESPS/general/src/tk_plot_tools/plotsd.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Joe Buck (originally)
 * Checked by:
 * Revised by:
 *
 * Brief description:
 *
 */


static char *sccs_id = "@(#)plotsd.c	1.2 7/23/98    ERL";
int	debug_level = 0;

/*----------------------------------------------------------------------+
|									|
+-----------------------------------------------------------------------+
|									|
|  plotsd -- plot ESPS sample data in a form suitable for plotas	|
|									|
|  Joseph T. Buck, EPI							|
|  adapted from a program by Shankar Narayan.				|
|  Converted 1986 Apr 22 by Rod Johnson to read new SPS format.		|
|  Modified for SPS Common by Ajaipal S. Virdy on 9/2/86		|
|  Modified for -m, -y, and -Y options by John Shore on 9/19/86		|
|  Modified for -E option by Rod Johnson Sept 1987.			|
|  Modified for FEA_SD by John Shore, Oct. 1989                         |
|									|
+----------------------------------------------------------------------*/

#include <stdio.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/unix.h>

#define PROGNAME "plotsd"
#define SYNTAX \
USAGE("plotsd [-{prs} range] [-t text]... [-x level] [-y range]\n\
 [-E expansion] [-M maxpoints] [-Y range] file")
#define TMPNAME "plsdXXXXXX"
#define BUFSIZE 1000
#define MAXTEXT 10
#define Printf (void) printf

void	plotscale(), plotexscale();
void	lrange_switch(), frange_switch();
char	*eopen();
char    *e_temp_name();

main(argc, argv)
    int     argc;
    char    **argv;
{
    char	*ProgName = PROGNAME;

    extern int	optind;		/* Used in command-line parsing by getopt */
    extern char	*optarg;	/* Used in command-line parsing by getopt */
    int		c;		/* Command-line option letter */

    long	nan;		/* Number of points in range to be plotted */
    long	ndrec;		/* Number of points in file */
    long	start_p,	/* First point in range to be plotted */
		last_p;		/* Last point in range to be plotted */
    long	tag0, tag1;	/* Endpoints of unexpanded range (E option) */
    int		xdp,		/* No. of decimal places for x scale marks */
		ydp;		/* No. of decimal places for y scale marks */
    long	i;		/* Loop index */
    int		ix, iy;		/* "Stanford" plot universe coordinates */
    long	numpoints;	/* Actual number of points plotted */
    long	maxpoints = 1000;   /* Approx. maximum no. of points to plot */
				  /* (if nan>maxpoints, points are skipped) */

    int		ntext = 0;	/* Number of lines of text below plot */
    int		pflag = NO,	/* -p option specified? */
		yflag = NO;	/* -y or -Y option specified? */
    int		yexact = 0;	/* -Y option specified? */
    double	expand = 0.0;	/* Expansion fraction specified with -E */
    int		Eflag = NO;	/* -E option specified? */
    char        *Wflag = NULL;  /* used to specify x geometry */
    char        *title = NULL;  /* used for holding user-specifed title -
				 not used yet here; see aplot.c */

    char	*srange = NULL;	/* Range (sec) specified with -s option */
    int		sflag = NO;	/* -s option specified? */

    char	*text[MAXTEXT];	/* Lines of text below plot */
    char	*prange = NULL;	/* Range (samples) specified with -p */
    char	*filename = NULL;   /* Input file name */

    float	*data;		/* Data to plot */
    int		interval;	/* Skip (interval - 1) pts. for each plotted */
    
    double	xmin, xmax;	/* Ends of x axis (data coordinates) */
    double	ymin , ymax;	/* Ends of y axis (data coordinates) */
    float	yminf, ymaxf;	/* Ends of y axis (data coordinates) */
    double	xstep, ystep;	/* Spacing between ticks on axes */
    int		nogrid=0;	/* suppress horizontal grid if 1 */

    FILE	*istrm;		/* Input file pointer */
    struct header   *h;		/* Input file header */
    int		Dflag=0;
    char	s[80];

/* parse command line for parameters */

    while ((c = getopt(argc, argv, "r:p:s:t:x:y:E:M:Y:D:W:")) != EOF)
	switch (c) 
        {
	    case 'r':
	    case 'p':
		prange = optarg;
		pflag++;
		break;
	    case 's':
		srange = optarg;
		sflag++;
		break;
	    case 't':
	        if (ntext < MAXTEXT)
		    text[ntext++] = optarg;
		else 
                {
		    Fprintf(stderr, "%s: Too many -t options\n", ProgName);
		    exit(1);
		}
		break;
	    case 'x':
		debug_level = atoi(optarg);
		break;
	    case 'y': 
		frange_switch(optarg, &ymin, &ymax);
		yflag++;
		break;
	    case 'E':
		expand = atof(optarg);
		Eflag = YES;
		break;
	    case 'M':
		maxpoints = atoi(optarg);
		break;
	    case 'Y': 
		frange_switch(optarg, &ymin, &ymax);
		yflag++;
		yexact = 1;
		break;
	    case 'D':
		Dflag = 1;
		break;
	    case 'W':
		Wflag = optarg;
	        break;
	    default:
		SYNTAX;
	}

/* Process input file name and read parameter file */
	
    if (argc != optind) filename = argv[optind];

    (void) read_params((char *) NULL, SC_CHECK_FILE, filename);

    if (filename == NULL)
    {
	if(symtype("filename") == ST_UNDEF)
	{
	    Fprintf(stderr, "%s: no input file\n", ProgName);
	    SYNTAX
	}
	filename = getsym_s("filename");
    }
     
    filename = eopen(ProgName, filename, "r", FT_FEA, FEA_SD, &h, &istrm);

    if (get_fea_siz("samples", h, (short) NULL, (long) NULL) != 1) {
      Fprintf(stderr, "%s: sorry, can't work on multi-channel files\n",
	      ProgName);
      exit(1);
    }

    tk_init(PROGNAME, filename, Wflag, title);

/* Find x-axis limits */

    if (pflag && sflag)
    {
	Fprintf(stderr,
	"%s: conflicting options: -p and -s\n", ProgName);
	exit(1);
    }

    start_p = 1;
    last_p = LONG_MAX;

    if (!pflag && !sflag)
    {
    	if (symtype("start") != ST_UNDEF) start_p = getsym_i("start");
    	if (symtype("nan") != ST_UNDEF) last_p = start_p + getsym_i("nan") - 1;
    }

    if (pflag) lrange_switch(prange, &start_p, &last_p, 0);

    if (sflag)
    {
	double	start_s,	    /* Beginning of range (sec) */
		last_s;		    /* End of range (sec) */

	/* Sampling frequency */
	double	sf = get_genhd_val("record_freq", h, 1.0);

	if (sf == 0.0) sf = 1.0;
	start_s = 0.0;
	last_s = FLT_MAX;
	frange_switch(srange, &start_s, &last_s);
	start_p = start_s * sf + 1;
	if (last_s == FLT_MAX)
	    last_p = LONG_MAX;
	else
	    last_p = last_s * sf + 1;
    }

    if (h->common.ndrec != -1)	/* Input is file. */
	ndrec = h->common.ndrec;    /*  Get ndrec from header. */
    else			/* Input is pipe. */
    {				    /* Use temp file to count. */
	char		*template = TMPNAME;
	char		*tempname;
	FILE		*tmpstrm;
	static float	buf[BUFSIZE];
	int		num_read;

	tempname = e_temp_name(template);
	TRYOPEN(ProgName, tempname, "w+", tmpstrm);
	(void) unlink(tempname); 
	ndrec = 0;
	do
	{
	    num_read = get_sd_recf(buf, BUFSIZE, h, istrm);
	    if (num_read != 0) put_sd_recf(buf, num_read, h, tmpstrm);
	    ndrec += num_read;
	} while (num_read == BUFSIZE);
	(void) rewind(tmpstrm);
	Fclose(istrm);
	istrm = tmpstrm;
    }

    if (last_p > ndrec) last_p = ndrec;

    if (start_p >= last_p)
    {
	 Fprintf(stderr, "%s: start point after end point!\n", ProgName);
	 exit(1);
    }

    nan = last_p - start_p + 1;

/* Write to ESPS common file before -E alters limits. */

    (void) putsym_s("filename", filename);
    (void) putsym_s("prog", ProgName);
    (void) putsym_i("start", (int) start_p);
    (void) putsym_i("nan", (int) nan);
    
    if (Eflag)			/* Expand range (-E option) */
    {
	tag0 = start_p; tag1 = last_p;
	if (expand < 0.0) expand = 0.0;
	    
	expand *= 0.5 * nan;
	plotscale((double) start_p - expand, (double) last_p + expand,
			1.0, &xmin, &xmax, &xstep, &xdp);
	if (xmin > 1)
	    start_p = LROUND(xmin);
	else
	    start_p = 1;
	if (xmax < ndrec)
	    last_p = LROUND(xmax);
	else
	    last_p = ndrec;
	nan = last_p - start_p + 1;
    }
    else
	plotscale((double) start_p, (double) last_p, 1.0, &xmin, &xmax, &xstep, &xdp);

    if (debug_level)
	Fprintf(stderr,
	"%s: start_p = %ld, nan = %ld, last_p = %ld, filename = %s\n",
	ProgName, start_p, nan, last_p, filename);

/*
 * Whew!
 * Begin program:
 *
 */

    if (debug_level)
	Fprintf(stderr, "%s: wrote all symbols\n", ProgName);

    if (last_p > ndrec) 
    {
	Fprintf(stderr, "%s: only %d points in file\n", ProgName, ndrec);
	exit(1);
    }

/* Read data. */

    skiprec(istrm, (long) start_p - 1, size_rec(h));

    if ( (data = (float *) malloc((unsigned) nan * sizeof(float))) == NULL )
    {
	Fprintf(stderr, "%s: can't allocate memory for %d data points\n",
		 ProgName, nan);
	exit(1);
    }

    if (debug_level)
	Fprintf(stderr, "%s: getting sampled data record\n", ProgName);

    (void) get_sd_recf(data, (int) nan, h, istrm);

    if (nan < maxpoints || maxpoints == 0)  /* Plot all pts. in range */
    {
	interval = 1;
	numpoints = nan;
    }
    else			/* Skip points */
    {
	interval = nan / maxpoints;
        numpoints = nan / interval;
    }

/* Get y-axis limits. */

    if (!yflag)
    {   
	yminf = ymaxf = (nan == 0) ? 0.0 : data[0];
        for (i = 1; i < nan; i++) 
	{
	    if (data[i] < yminf)
		yminf = data[i];
	    else if (data[i] > ymaxf)
		ymaxf = data[i];
	}
	ymin = (double) yminf;
	ymax = (double) ymaxf;
	if ((ymax - ymin) < .001) {
		ymax+=.001;
		nogrid = 1;
	}
    }



    if (debug_level)
	Fprintf(stderr, "%s: plotting points\n", ProgName);

    if (yexact) 
	plotexscale(ymin, ymax, 1.0, &ymin, &ymax, &ystep, &ydp);
    else
	plotscale(ymin, ymax, 1.0, &ymin, &ymax, &ystep, &ydp);

    draw_box(xmin, xmax, xstep, xdp, ymin, ymax, ystep, ydp, nogrid);
    tk_snd_plot_cmd("c 2");
    plot_dataf(data, numpoints, xmin, xmax, ymin, ymax, start_p, interval);
    tk_snd_plot_cmd("c 5");
    tk_snd_plot_cmd("m 3400 500");
    tk_snd_plot_cmd("t 5 1");
    sprintf(s,"File: %s",filename);
    tk_snd_plot_cmd(s);

    if (Eflag)
    {
	iy = 3000 + ymin * 2500.0 / (ymax - ymin);

	ix = 500.0 + (tag0 - xmin) * 5000.0 / (xmax - xmin);
	sprintf(s,"m %d %d", iy, ix);
	tk_snd_plot_cmd(s);
	sprintf(s,"d 500 %d", ix);
	tk_snd_plot_cmd(s);
	sprintf(s,"m 480 %d",ix);
        tk_snd_plot_cmd("t 4 2");
	sprintf(s,"%d", tag0);
        tk_snd_plot_cmd(s);

	ix = 500.0 + (tag1 - xmin) * 5000.0 / (xmax - xmin);
	sprintf(s,"m %d %d", iy, ix);
	tk_snd_plot_cmd(s);
	sprintf(s,"d 500 %d", ix);
	tk_snd_plot_cmd(s);
	sprintf(s,"m 480 %d",ix);
        tk_snd_plot_cmd(s);
        tk_snd_plot_cmd("t 4 2");
        sprintf(s,"%d", tag1);
	tk_snd_plot_cmd(s);
    }

    iy = 3600;
    for (i = 0; i < ntext; i++) 
    {
	tk_snd_plot_cmd("c 5");
        tk_snd_plot_cmd_1arg("m %d 500", iy);
	tk_snd_plot_cmd("t 5 1");
        tk_snd_plot_cmd(text[i]);
	iy += 200;
    }

    if(!Dflag) /* supress date */
    	print_time(200, 4400);

    tk_start();
    
    (void) putsym_i("beginplot", (int) xmin);
    (void) putsym_i("endplot", (int) xmax);

    exit(0);
    /*NOTREACHED*/
}

